File: 210.Course-Schedule-II.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function (numCourses, prerequisites) {
  const preReqMap = new Map();
  const visiting = new Set();
  const visited = new Set();
  const result = [];

  for (let i = 0; i < numCourses; i++) {
    preReqMap.set(i, []);
  }

  for (const [course, prereq] of prerequisites) {
    preReqMap.set(course, [...preReqMap.get(course), prereq]);
  }

  const dfs = (node) => {
    if (visited.has(node)) {
      return true;
    }

    if (visiting.has(node)) {
      return false;
    }

    visiting.add(node);

    const prereqs = preReqMap.get(node);

    for (const prereq of prereqs) {
      if (!dfs(prereq)) {
        return false;
      }
    }

    visiting.delete(node);
    visited.add(node);
    result.push(node);

    return true;
  };

  for (let i = 0; i < numCourses; i++) {
    if (!dfs(i)) {
      return [];
    }
  }

  return result;
};

File: python/210.Course-Schedule-II.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        result = []
        visited = set()
        visiting = set()
        prereq_map = {}

        for i in range(numCourses):
            prereq_map[i] = []

        for course, prereq in prerequisites:
            prereq_map[course].append(prereq)

        def dfs(course):
            if course in visited:
                return True

            if course in visiting:
                return False

            visiting.add(course)

            for prereq in prereq_map[course]:
                if not dfs(prereq):
                    return False
            
            result.append(course)
            visiting.remove(course)
            visited.add(course)

            return True

        for i in range(numCourses):
            if not dfs(i):
                return [] 

        return result